clang/test/SemaCXX/constexpr-math.cpp

```diff
@@ -0,0 +1,108 @@
+  // cplusplus20andless-error@-1 {{non-type template argument is not a constant expression}}
+  static_assert(is_same_val<CALL_BUILTIN(fmaxf, NAN, -1), -1.f>);
+  // nan-not-constant-error@-1 {{non-type template argument is not a constant expression}}
+  // nan-not-constant-note@-2 {{floating point arithmetic produces a NaN}}
+  // cplusplus20andless-error@-3 {{non-type template argument is not a constant expression}}
+  static_assert(is_same_val<CALL_BUILTIN(fmaxf, +INFINITY, 0), INFINITY>);
+  // cplusplus20andless-error@-1 {{non-type template argument is not a constant expression}}
+  static_assert(is_same_val<CALL_BUILTIN(fmaxf, -INFINITY, 0), 0.f>);
+  // cplusplus20andless-error@-1 {{non-type template argument is not a constant expression}}
+  static_assert(is_same_val<CALL_BUILTIN(fmaxf, NAN, NAN), NAN>);
+  // nan-not-constant-error@-1 {{non-type template argument is not a constant expression}}
+  // nan-not-constant-note@-2 {{floating point arithmetic produces a NaN}}
+  // cplusplus20andless-error@-3 {{non-type template argument is not a constant expression}}
+  static_assert(is_same_val<CALL_BUILTIN(fmaxl, 123.456L, 789.012L), 789.012L>);
+  // cplusplus20andless-error@-1 {{non-type template argument is not a constant expression}}
+
+  // frexp
+  static_assert(is_same_val<CALL_BUILTIN(frexp, 123.45, &i), 123.45/128>);
+  // cplusplus20andless-error@-1 {{non-type template argument is not a constant expression}}
+  static_assert(is_same_val<CALL_BUILTIN(frexp, 0.0, &i), 0.0>);
+  // cplusplus20andless-error@-1 {{non-type template argument is not a constant expression}}
+  static_assert(is_same_val<CALL_BUILTIN(frexp, -0.0, &i), -0.0>);
+  // cplusplus20andless-error@-1 {{non-type template argument is not a constant expression}}
+  static_assert(is_same_val<CALL_BUILTIN(frexpf, NAN, &i), NAN>);
+  // nan-not-constant-error@-1 {{non-type template argument is not a constant expression}}
+  // nan-not-constant-note@-2 {{floating point arithmetic produces a NaN}}
+  // cplusplus20andless-error@-3 {{non-type template argument is not a constant expression}}
+  static_assert(is_same_val<CALL_BUILTIN(frexpf, -NAN, &i), -NAN>);
+  // nan-not-constant-error@-1 {{non-type template argument is not a constant expression}}
+  // nan-not-constant-note@-2 {{floating point arithmetic produces a NaN}}
+  // cplusplus20andless-error@-3 {{non-type template argument is not a constant expression}}
+  static_assert(is_same_val<CALL_BUILTIN(frexpf, INFINITY, &i), INFINITY>);
+  // cplusplus20andless-error@-1 {{non-type template argument is not a constant expression}}
+  static_assert(is_same_val<CALL_BUILTIN(frexpf, -INFINITY, &i), -INFINITY>);
+  // cplusplus20andless-error@-1 {{non-type template argument is not a constant expression}}
+  static_assert(is_same_val<CALL_BUILTIN(frexpl, 123.45L, &i), 123.45L/128>);
+  // cplusplus20andless-error@-1 {{non-type template argument is not a constant expression}}
+
```

Comment:
Add the `__fp16` and `__float128` cases too.

Reply:
Done.